### ISSUE TITLE
Null modification with concat assignment

### ISSUE THREAD

**Issue by rentalhost** [kalessil/phpinspectionsea#374]
Check [**this code**](https://3v4l.org/bJF4m):

```php
function set(string $value)
{
    $value .= null;
}
```

In this case, `null` will not affect the `$value` type, it will keep being a `string`.

> PHP | Php Inspections (EA Extended) | Architecture | **Callable parameter usage violates definition**

**Comment by kalessil** [comment]
Appending null to string looks strange, what is the intention of this concatenation?

**Comment by rentalhost** [comment]
It is a reduced case. On my side I have a method that returns `?string`, and it causes this same issue.

**Comment by Fleshgrinder** [comment]
That method probably should return `string` only. Either the returned `null` has a special meaning and you should account for it, or it could be an empty string. I do not see a reason for special casing here.

**Comment by rentalhost** [comment]
It is not true. In my case I have a method that returns a description for an object. When this description is not available for the instance it returns NULL, that is valid for some kind of context. Then I pass this instance to another method that can use this description to append to a string builder. Even on this context this description is not required and it can be just ignored (but this string have another informations that comes before this description).

**Comment by kalessil** [comment]
I've dug deeper: the case seems to be totally valid - https://stackoverflow.com/a/32284602
WIll check why it reported.

**Comment by Fleshgrinder** [comment]
@kalessil this is not about whether it is a problem or not in regards to errors. Applying that logic would also mean that `int` in combination with `null` is valid, so is `float`, `bool`, … because of PHP’s coercion stuff. The intention of the `null` warnings is to make developers aware of mixing this kind of stuff.

An explicit cast should be added by a developer to disable the warning (and the extension should account for that by removing the warning). This shows that the developer knows what she’s doing, instead of relying on implicit rules.

**Comment by rentalhost** [comment]
Cast is valid, but is an additional op that is not required here. 

The `.=` is a string concat operation. Then it should expects a `string|null`, but not `int|null` (altrough it is valid too) because it can turn the code confusing (in this case, cast should be applied, then user will garantee that is the expected case: he knows that is receiving an int but need cast to string to make compatible with the variable type).

**Comment by Fleshgrinder** [comment]
Casting on your own is most probably faster than letting the engine figuring it out. This would actually require benchmarking, but the operation is always there, regardless of you hiding it or not. This is exactly how every strict language handles this and this feature was meant to bring this strictness to PHP. You are watering the whole thing by adding special cases to it, and make it harder to understand for everyone. This is a bad idea.

**Comment by kalessil** [comment]
Fixed: it was a bug in my implementation, type inference working well.

**Comment by rentalhost** [comment]
@Fleshgrinder For PHP it doesn't matter. `null` is converted to an empty string if you cast it to string (that happen on `.=` operator). Then it is not a bug and is totally previsible.
